go/felix/ifacemonitor/ifacemonitor_test.go

```diff
@@ -0,0 +1,300 @@
+	link := nl.links[name]
+	link.addrs.Add(addr)
+	nl.links[name] = link
+	nl.signalAddr(name, addr, true)
+}
+
+func (nl *netlinkTest) delAddr(name string, addr string) {
+	link := nl.links[name]
+	link.addrs.Discard(addr)
+	nl.links[name] = link
+	nl.signalAddr(name, addr, false)
+}
+
+func (nl *netlinkTest) signalAddr(name string, addr string, exists bool) {
+	// Build the update.
+	net, err := netlink.ParseIPNet(addr)
+	if err != nil {
+		panic("Address parsing failed")
+	}
+	update := netlink.AddrUpdate{
+		LinkIndex:   nl.links[name].index,
+		NewAddr:     exists,
+		LinkAddress: *net,
+	}
+
+	// Send it.
+	log.Info("Test code signaling an addr update")
+	nl.addrUpdates <- update
+}
+
+func (nl *netlinkTest) Subscribe(
+	linkUpdates chan netlink.LinkUpdate,
+	addrUpdates chan netlink.AddrUpdate,
+) error {
+	nl.linkUpdates = linkUpdates
+	nl.addrUpdates = addrUpdates
+	return nil
+}
+
+func (nl *netlinkTest) LinkList() ([]netlink.Link, error) {
+	links := []netlink.Link{}
+	for name, link := range nl.links {
+		var rawFlags uint32 = 0
+		if link.state == "up" {
+			rawFlags = syscall.IFF_RUNNING
+		}
+		links = append(links, &netlink.Dummy{
+			LinkAttrs: netlink.LinkAttrs{
+				Name:     name,
+				Index:    link.index,
+				RawFlags: rawFlags,
+			},
+		})
+	}
+	return links, nil
+}
+
+func (nl *netlinkTest) AddrList(link netlink.Link, family int) ([]netlink.Addr, error) {
+	name := link.Attrs().Name
+	model, prs := nl.links[name]
+	addrs := []netlink.Addr{}
+	if prs {
+		model.addrs.Iter(func(item interface{}) error {
+			addr := item.(string)
+			net, err := netlink.ParseIPNet(addr)
+			if err != nil {
+				panic("Address parsing failed")
+			}
+			if strings.ContainsRune(addr, ':') {
+				if family == netlink.FAMILY_V6 {
+					addrs = append(addrs, netlink.Addr{
+						IPNet: net,
+					})
+				}
+			} else {
+				if family == netlink.FAMILY_V4 {
+					addrs = append(addrs, netlink.Addr{
+						IPNet: net,
+					})
+				}
+			}
+			return nil
+		})
+	}
+	return addrs, nil
+}
+
+var linkC = make(chan string, 1)
+
+func linkStateCallback(ifaceName string, ifaceState ifacemonitor.State) {
+	log.Info("linkStateCallback: ifaceName=", ifaceName)
+	log.Info("linkStateCallback: ifaceState=", ifaceState)
+	linkC <- ifaceName
+}
+
+func expectLinkStateCb(ifaceName string) {
+	cbIface := <-linkC
+	Expect(cbIface).To(Equal(ifaceName))
+}
+
+var addrC = make(chan string, 1)
+
+func addrStateCallback(ifaceName string, addrs set.Set) {
+	log.Info("addrStateCallback: ifaceName=", ifaceName)
+	log.Info("addrStateCallback: addrs=", addrs)
+	addrC <- ifaceName
+}
+
+func expectAddrStateCb(ifaceName string) {
+	cbIface := <-addrC
+	Expect(cbIface).To(Equal(ifaceName))
+}
+
+var _ = Describe("ifacemonitor", func() {
+	It("New", func() {
+
+		// Make an Interface Monitor that uses a test netlink
+		// stub implementation and resync trigger channel -
+		// both controlled by this code.
+		nl := &netlinkTest{}
```

Comment:
I've done the BeforeEach, but don't think we should do the goroutine stopping.  There is no way of doing this cleanly unless we change the code under test, and then there would be a risk of missing - during real operation - that the interface monitor goroutine had exited for some reason.  I don't think it's worth the risk, for the sake of a goroutine clean up that isn't actually an issue as the UT currently stands.